h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -0,0 +1,195 @@
+        nodesToTraverse.add(sharedTreeSubgraph.rootNode);
+        append(treeprops._rightChildren, treeprops._leftChildren,
+                treeprops._descriptions, treeprops._thresholds, treeprops.features, treeprops.nas,
+                nodesToTraverse, -1, false);
+
+        return treeprops;
+    }
+
+    private static void append(final int[] rightChildren, final int[] leftChildren, final String[] nodesDescriptions,
+                               final float[] thresholds, final String[] splitColumns, final boolean[] naHandlings,
+                               List<SharedTreeNode> nodesToTraverse, int pointer, boolean visitedRoot) {
+        if(nodesToTraverse.isEmpty()) return;
+
+        List<SharedTreeNode> discoveredNodes = new ArrayList<>();
+
+        for (SharedTreeNode node : nodesToTraverse) {
+            pointer++;
+            final SharedTreeNode leftChild = node.getLeftChild();
+            final SharedTreeNode rightChild = node.getRightChild();
+            if(visitedRoot){
+                fillnodeDescriptions(node, nodesDescriptions, thresholds, splitColumns, naHandlings, pointer);
+            } else {
+                nodesDescriptions[pointer] = "Root node";
+                visitedRoot = true;
+            }
+
+            if (leftChild != null) {
+                discoveredNodes.add(leftChild);
+                leftChildren[pointer] = leftChild.getNodeNumber();
+            } else {
+                leftChildren[pointer] = NO_CHILD;
+            }
+
+            if (rightChild != null) {
+                discoveredNodes.add(rightChild);
+                rightChildren[pointer] = rightChild.getNodeNumber();
+            } else {
+                rightChildren[pointer] = NO_CHILD;
+            }
+        }
+
+        append(rightChildren, leftChildren, nodesDescriptions, thresholds, splitColumns, naHandlings,
+                discoveredNodes, pointer, true);
+    }
+
+    private static void fillnodeDescriptions(final SharedTreeNode node, final String[] nodeDescriptions,
+                                             final float[] thresholds, final String[] splitColumns,
+                                             final boolean[] naHandlings, final int pointer) {
+        final StringBuilder nodeDescriptionBuilder = new StringBuilder();
+
+        if (!Float.isNaN(node.getParent().getSplitValue())) {
+            nodeDescriptionBuilder.append(node.getParent().getColName());
+            if (node.isLeftward()) {
+                nodeDescriptionBuilder.append(" < ");
+            } else {
+                nodeDescriptionBuilder.append(" >= ");
+            }
+            nodeDescriptionBuilder.append(node.getParent().getSplitValue());
+            thresholds[pointer] = node.getParent().getSplitValue();
+        } else {
+            final BitSet childInclusiveLevels = node.getInclusiveLevels();
+            final int cardinality = childInclusiveLevels.cardinality();
+            if ((cardinality > 0)) {
```

Comment:
unnecessary double parenthesis (and I love Lisp :))

Reply:
Removed.